### ISSUE TITLE
Update libseccomp without imposing Linux 5.7 requirement (seccomp API level 6)

### ISSUE THREAD

**Issue by Furisto** [gitpod-io/gitpod#7793]
### Bug description

We encountered an issue in staging and preview environments where workspaces could not start up successfully. This issue popped up after #7657 was introduced. The change was reverted [here](https://github.com/gitpod-io/gitpod/pull/7714). This only happened in the preview and staging environments which are running kernel version 5.4, but not in production which is running 5.13. 

Kernel 5.4 has libseccomp api level 5 while 5.13 has api level 6. The first kernel version to support api level 6 is 5.7. A kernel has supports libseccomp api level 6 if it [supports](https://github.com/seccomp/libseccomp/blob/51b50f95e1fb717e4560818f8b90b7ebde314ad1/src/api.c#L160-L196) SECCOMP_FILTER_FLAG_TSYNC_ESRCH.
```
{"level":"info","message":"signaling to child process","ring":1,"serviceContext":{"service":"workspacekit","version":"commit-c9cebdd0ef6cd7cc09456a24a48afd4ebe82d085"},"severity":"INFO","time":"2022-01-20T10:11:12Z"}
{"level":"info","message":"awaiting seccomp fd","ring":1,"serviceContext":{"service":"workspacekit","version":"commit-c9cebdd0ef6cd7cc09456a24a48afd4ebe82d085"},"severity":"INFO","time":"2022-01-20T10:11:12Z"}
{"@type":"type.googleapis.com/google.devtools.clouderrorreporting.v1beta1.ReportedErrorEvent","error":"cannot add rule for bind:\n    github.com/gitpod-io/gitpod/workspacekit/pkg/seccomp.LoadFilter\n        /tmp/build/components-workspacekit--app.e12d4c33816c4ea1bcaadffc719ca0288d37780e/pkg/seccomp/notify.go:86\n  - two checks on same syscall argument","level":"error","message":"cannot load seccomp filter - syscall handling would be broken","ring":2,"serviceContext":{"service":"workspacekit","version":"commit-c9cebdd0ef6cd7cc09456a24a48afd4ebe82d085"},"severity":"ERROR","time":"2022-01-20T10:11:12Z"}
```

### Steps to reproduce

Apply #7657  to kernel < 5.7

### Workspace affected

n.a.

### Expected behavior

Workspaces starting successfully.

### Example repository

n.a.

### Anything else?

See https://gitpod.slack.com/archives/C01KGM9AW4W/p1642679350077600

### How to test

Short term
1. Make sure to test in core-dev (GKE & K8s) and workspace-preview (K3s).

Long term
1. Once core-dev is gone, we'll need to change the kernel version in workspace-preview, if we are to test against a kernel that is like the one used in GKE.

```shell
echo "deb http://security.ubuntu.com/ubuntu focal-security main " >>/etc/apt/sources.list
linux-gke linux-gke-headers-5.4.0-1059 linux-gke-tools-5.4.0-1059 linux-headers-5.4.0-1059-gke linux-image-5.4.0-1059-gke linux-image-gke linux-modules-5.4.0-1059-gke linux-modules-extra-5.4.0-1059-gke linux-tools-5.4.0-1059-gke
```

**Comment by utam0k** [comment]
## Why do we need to use `kinvolk/libseccomp-golang`?

This is because we want to use the seccomp notify feature. We use seccomp notify when the proc mount system call, we make the target path change in userspace. So if seccomp notify doesn't work properly, we can't mount the proc.
https://github.com/gitpod-io/gitpod/blob/main/components/workspacekit/pkg/seccomp/notify.go

## Why do we need to upgrade the version of `kinvolk/libseccomp-golang`?

The libseccomp-go that we are currently using has a bug regarding seccomp notify. It is probably fixed in this commit.
https://github.com/kinvolk/libseccomp-golang/commit/3879420cc921efb4f1a2d75bea84e6158ef21a78

Normally, we should re-execute the corresponding system call when we get an error code of `EINTR`, but that process was missing in libseccomp-go. The above commit is the fix. Probably because of this bug, seccomp notify was not working properly, so the proc mount was not working properly.

https://man7.org/linux/man-pages/man7/signal.7.html
> In certain circumstances, the seccomp(2) user-space notification feature can lead to restarting of system calls that would otherwise never be restarted by SA_RESTART; for details, see  seccomp_unotify(2).

https://man7.org/linux/man-pages/man3/errno.3.html

## Why can't we update `kinvolk/libseccomp-golang`?

This is because the kernel version needs to be upgraded as outlined in the issue.

## What should we do?

So, we can fork `kinvolk/libseccomp-golang` and apply the above commit to the version we want to use, to avoid this problem. This way, we won't get caught up in the seccomp API level-up.
https://github.com/gitpod-io/libseccomp-golang

**Comment by Furisto** [comment]
- Applying commit [kinvolk/libseccomp-golang@3879420 ](https://github.com/kinvolk/libseccomp-golang/commit/3879420cc921efb4f1a2d75bea84e6158ef21a78) won't do as any good as we are using a development version of the kinvolk library which [already has these changes.](https://github.com/kinvolk/libseccomp-golang/commit/88391784331306c118f447d44ad9c259dcf9e130)
- Using the latest version from the main branch also does not fix the issue
- Switching the shift method from fuse to shifts however does work

**Comment by kylos101** [comment]
Hey @utam0k @csweichel , 

@Furisto and I spoke about this issue today, and after discussing the current state, decided to move this back to scheduled for now. 

@Furisto may I ask you to add a note, to help us pick this up at a later time?

**Comment by kylos101** [comment]
> Hey @utam0k @csweichel ,
> 
> @Furisto and I spoke about this issue today, and after discussing the current state, decided to move this back to scheduled for now.
> 
> @Furisto may I ask you to add a note, to help us pick this up at a later time?

@Furisto I was wrong. :smile: 

May we ask you to create an issue with the kind folks @ Kinvolk, to see if they can help point us in the right direction? 

Once that issue has been created, share it here, and we'll use that as reference if we have to reach out via other channels (Twitter, etc.).

**Comment by kylos101** [comment]
Reassigning @Furisto (as he'll be creating an issue, reaching out to Kinvolk for guidance).

**Comment by utam0k** [comment]
@kylos101 Thanks for your help 🙏  I have some I found a curious thing. The commit hash we are using seems to have been erased from main (or was never in main in the first place).
https://github.com/seccomp/libseccomp-golang/commits/88391784331306c118f447d44ad9c259dcf9e130

Also, it seems that seccomp notify is already implemented in `seccomp/libseccomp-golang` under fork. We may need to figure out the issue more precisely before we can create an issue. I feel like we need to find a way to reproduce it minimally for example. What do you think?
I'm sorry if you have already decided what issue you are going to make in the discussion.

**Comment by utam0k** [comment]
related commit: https://github.com/gitpod-io/gitpod/commit/5f9f701835292bbe28e71c8eb70f8a09965cb5ab

**Comment by utam0k** [comment]
Seeing that there is no problem with shiftfs, there may be a problem with the combination of fusefs and seccomp notify.
For example, there was an issue with lxd like this.
https://github.com/lxc/lxd/issues/6440

**Comment by utam0k** [comment]
I found it! Thanks. 
> The development branch has the commit you are talking about.

**Comment by Furisto** [comment]
@utam0k I tried to switch to the official seccomp golang lib (see [here](https://github.com/gitpod-io/gitpod/tree/furisto/seccomp)) but this causes workspaces not to open for me for in the preview environment (saying it has not found a backup for the workspace)

Another option could be to set our seccomp profile to allow everything and see if that improves anything. Could be that fuse does syscalls that are blocked by seccomp.

**Comment by utam0k** [comment]
@Furisto 
I tried to find out the cause of the error message. The reason why you get this message is because of a mistake in the Rule configuration. I think clarifying the cause of this error message would be a good next step.
https://github.com/seccomp/libseccomp-golang/blob/2a71848205430267e34457b3b771a37e7d584956/seccomp_internal.go#L433

Somewhere in `seccomp_rule_add_array()`, EINVAL should be returned.
https://github.com/seccomp/libseccomp/blob/3f0e47fe2717b73ccef68ca18f9f7297ee73ebb2/src/api.c#L541-L565
https://github.com/seccomp/libseccomp/blob/c305ef351bc9b254f3f65359ca2e6435bb920752/src/db.c#L1192-L1208

Starting from linux kernel 5.7, it is possible to set both TSYNC and USER_NOTIF at the same time. Maybe this is the cause, but I could not be sure today.
https://github.com/torvalds/linux/commit/51891498f2da78ee64dfad88fa53c9e85fb50abf

By the way, I was working on this commit now.
https://github.com/gitpod-io/gitpod/commit/e3919777c4a960d39d549e523e64a3045c01f811

**Comment by utam0k** [comment]
I have created a repository to reproduce libseccomp errors. We can reproduce it by changing the URL to the preview environment when you open it. This will speed up the feedback loop so that we can try more things ;)
https://github.com/utam0k/libseccomp-golang-playground

**Comment by Furisto** [comment]
@utam0k Found out what the issue with the official libseccomp golang is. [This](https://github.com/seccomp/libseccomp/blob/51b50f95e1fb717e4560818f8b90b7ebde314ad1/src/db.c#L1200-L1202) is the place where the error occurs in libseccomp. It is not possible to use Tsync and Notify together unless we are api level 6. [Here](https://github.com/seccomp/libseccomp-golang/blob/2a71848205430267e34457b3b771a37e7d584956/seccomp.go#L620-L625) Tsync is set by libseccomp-golang. When I commented out this part adding the rule was successful.

**Comment by utam0k** [comment]
@Furisto Thanks for your report. It is very helpful to me.
We finally succeeded in getting the forked repository from seccomp/libseccomp to work with linux kernel 5.6 and below, API level 6 and below 🎉 
I've sent out a PR for the changes I need to make to the forked repository.
https://github.com/gitpod-io/libseccomp-golang/pull/1
After this PR is merged, I will make a PR to change workspacekit to use `gitpod-io/libseccomp-golang`.
This branch is already running libseccomp-golang, an improved mainstream version.
https://github.com/gitpod-io/gitpod/tree/to/libseccomp

**Comment by utam0k** [comment]
@kylos101 @Furisto 
docker did not work with fusefs in the upgraded environment of libseccomp. This reveals the next problem. The next step is to investigate and resolve the behavior of fusefs and seccomp notify (and possibly other factors). However, this is not the purpose of this issue and I think it should be a different issue. What do you think?

I did some research on this today and all I know is that the target of moveMount() does not exist in the mount system call (type = proc) hooked by seccomp notify. I guess the next step is to plant a log, find a small way to reproduce it and speed up the feedback loop.

You can reproduce the error in the preview environment that reverts this commit.
https://github.com/gitpod-io/gitpod/commit/5f9f701835292bbe28e71c8eb70f8a09965cb5ab

**Comment by Furisto** [comment]
@utam0k I created new issue: #8020
